### ISSUE TITLE
max_cat_threshold default = 32 with large number of categories

### ISSUE THREAD

**Issue by memeplex** [microsoft/LightGBM#5516]
In https://github.com/microsoft/LightGBM/issues/2261 it was said that

>> Then because max_cat_threshold = 1, the algorithm only has one split point to evaluate and this split point is as close to the median (as determined by number of observations or weighted observations) as possible?

>  I want to clarify that, the split point will be the one with the highest (or lowest) sum(gradients) / sum(hessians) 

Now, by looking at the code in https://github.com/microsoft/LightGBM/blob/346f88398282c5677dcaa880e147026eadba29e4/src/treelearner/feature_histogram.hpp#L397

the above assertion seems dubious to me. If max_cat_threshold is 1 then max_num_cat is 1 and the main loop will only consider one split at all! Of course it will be the optimal split between all considered, in case there actually is a valid split, since it's the only one.

The documentation of the parameter is not crystal clear either:

> limit number of split points considered for categorical features. 

Now, what concerns me is that the default value is 32. What would happen when there is a large number of categories for a feature? Only the first 32 split points will be considered? I may be missing something, but that seems like a dangerous default to me.

@guolinke what do you think?

**Comment by guolinke** [comment]
@memeplex this is mainly for regularization purposes. with too many categories threshold, it is easily overfitting. 
A small set of categories is better in generalization.

**Comment by memeplex** [comment]
I understand that, but one thing is to do a few significant splits and another things is to look at a few candidate splits. My concern is that by looking at just the first 32 candidates for a feature with hundreds of categories you may end up never splitting that feature even when there are very significant splits after the 32 inspected ones. In terms of the default histogram for such a feature, what percentage of the number of categories would 32 be?

**Comment by guolinke** [comment]
@memeplex the categorical split points are sorted by split gain, so the significant splits will be used first.
